### ISSUE TITLE
Optimisers/Convenient implementation for popular Bayesian learning methods

### ISSUE THREAD

**Issue by jingmiao-g** [pytorch/pytorch#106189]
### 🚀 The feature, motivation and pitch
Hello,

I am trying to do Bayeisan inference for neural network parameters.
This work usually requires adding a random noise to model weights/parameters in each updates.
I tried in pytorch using the following method:
#-----------------------------------
with torch.no_grad():
        for p in model.parameters():
            p.add_(a*torch.randn(p.size()))
#---------------------------------
However, I found that the result is not correct (I have ground truth calculated by julia), I suspect that this is because the added noise will NOT be tracked during training. The bayesian inference I've tried then does not converge at a proper level in pytorch.

Is there anyway to realise it or is it possible to add a new feature to pytorch? Most state-of-the-art systems are implemented by pytorch and so it would be more convenient for those who are doing research in Bayesian learning.

The usual bayesian learning methods include stochastic gradeint langevin dynamics(SGLD), Hamiltonian Monte-Carlo(HMC), stochastic gradient HMC, variational inference (Bayes-by-backprop), etc.

Thank you.

### Alternatives

Julia package Flux.jl can be used for such task easily but most SOTA systems are in pytorch.
Flux.jl uses Zygote for gradient calculations, where the network parameters can be modified manually in a more convenient style.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @vincentqb @janeyx99 @crcrpar

**Comment by janeyx99** [comment]
Yea, with the torch.no_grad(), the computation within the context will not be backprop'd over. 

I am curious if you can use our forward hooks to achieve something similar with the parameters and get tracking. https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html

Let me know if you have questions or if I did not correctly understand your feature ask!

**Comment by janeyx99** [comment]
To be clear, I can imagine something like:
```
def hook(module, input, output) -> None:
    for p in model.parameters():
        p.add_(a*torch.randn(p.size()))

model = ...<insert your module>
model.register_module_forward_hook(hook)
```
Note that my proposal would not change the input or output of the forward(), but would modify the parameters for the next iteration. If you wanted to modify the output, you would be able to do so as well and optionally return it from the hook.

**Comment by jingmiao-g** [comment]
Hi, 
Thanks for the reply and the proposed method.
I am trying to migrate from julia to pytorch and also finding if there is something I'd missed.
I got several tasks which require: 
#-------------------
p.add_(a*torch.randn(p.size()))
#-------------------
Last time I got different loss behaviors when using different languages. 
I now get one of the tasks correct so the code does noise the weights correctly and the next iteration will work with the noised weights rather its original.
I will see if there are differences between julia and torch on other tasks. Now I close this issue.
